src/VisualStudio/Core/Def/Implementation/Venus/ContainedDocument.DocumentServiceProvider.cs

```diff
@@ -0,0 +1,85 @@
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Text.Projection;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Venus
+{
+    internal sealed partial class ContainedDocument
+    {
+        public class DocumentServiceProvider : IDocumentServiceProvider
+        {
+            public static readonly IDocumentServiceProvider Instace = new DocumentServiceProvider();
+
+            private readonly SpanMapper _service;
+
+            private DocumentServiceProvider()
+            {
+                _service = new SpanMapper();
+            }
+
+            public TService GetService<TService>() where TService : IDocumentService
+            {
+                if (_service is TService service)
+                {
+                    return service;
+                }
+
+                return default;
+            }
+
+            private class SpanMapper : ISpanMapperService
+            {
+                public async Task<ImmutableArray<DocumentSpan>> MapSpansAsync(Document document, IEnumerable<TextSpan> spans, CancellationToken cancellationToken)
+                {
+                    var containedDocument = ContainedDocument.TryGetContainedDocument(document.Id);
+                    if (containedDocument == null)
+                    {
+                        return spans.Select(s => new DocumentSpan(document, s)).ToImmutableArray();
+                    }
+
+                    var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                    var editorSnapshot = sourceText.FindCorrespondingEditorTextSnapshot() as IProjectionSnapshot;
+                    if (editorSnapshot == null)
+                    {
+                        return spans.Select(s => new DocumentSpan(document, s)).ToImmutableArray();
+                    }
+
+                    var builder = ArrayBuilder<DocumentSpan>.GetInstance();
+                    foreach (var span in spans)
+                    {
+                        foreach (var primarySpan in editorSnapshot.MapToSourceSnapshots(span.ToSpan()))
+                        {
+                            // this is from http://index/?query=MapSecondaryToPrimarySpan&rightProject=Microsoft.VisualStudio.Editor.Implementation&file=VsTextBufferCoordinatorAdapter.cs&line=177
+                            // make sure we only consider one on primary buffer
+                            if (containedDocument.DataBuffer != primarySpan.Snapshot.TextBuffer)
+                            {
+                                continue;
+                            }
+
+                            // and consider span that is not split
+                            if (primarySpan.Length != span.Length)
+                            {
+                                continue;
```

Comment:
This would break the contract that MapSpansAsync would return a 1:1 for the original span going in and the resulting mapped span. In other words, if you put four spans in and get three out, you won't even know which of the four failed.

Reply:
sure. I can keep the 1:1 mapping. either Document being null or null entry.